### DISCUSSION TITLE
Catching Events in Tasmota Script - Howto?

### DISCUSSION THREAD

**Discussion by forumlw06** [arendst/Tasmota#20684] [Q&A] [answered]
Hello, 

i have 2 Tasmota devices (both on 13.3.0). 

Device1 is standard Tasmota.bin
Device2 is 13.3.0 with script and sendmail enabled (sending mails from a script works from that device)

What i want to achieve: 

- from Device1 i want so send events to device2 using `WebSend [Device2] event evt=1` (triggered by rules)
- On Device2 i want to catch that event "evt=1" and send mails with different content (depending on the sent event)

I read that i can catch events in Tasmota scripts in the `>E` section, but i am unsuccessful in doing so.

On Device2 i have this script:
```
>D

>E
print ===event ===
```

Every time i issue the Websend command on Device1 i see on Device2 the "===event===" message in the console.

**But how can i find out, what event was raised, so that i can react accordingly?**

I would like to have something like that, but not sure how to code that (CODE NOT WORKING, just to demonstrate, what i want to achieve):
```
>D

>E
print ===event ===
if (evt=1) {
=>sendmail [*:*:*:*:*:<to.mail@web.de>:Event] evt=1
}
if (evt=2) {
=>sendmail [*:*:*:*:*:<to.mail@web.de>:Event] evt=2
}
```

Thanks for your help. I saw in the docs, that `>E` is triggerd on RSL messages. But how to find out, what really is sent to `>E`.
Is there a way to print the raw data that is coming in?

Sorry for the newbie question.

P.S.: I want this kind of setup to be able to run my Tasmota plugs mainly on standard tasmota, but have the possibility to send mails on different situations. So setting up one ESP32 with sendmail enabled as "central mail relay", while most other devices can stay on standard tasmota

**Comment by gemu2015** [comment]
simple set a variable via websend
like:
websend [192.168.188.76] script>evt=1

```
>D
evt=0
>E
if (evt==1) {
=>sendmail [*:*:*:*:*:<to.mail@web.de>:Event] evt=1
}
if (evt==2) {
=>sendmail [*:*:*:*:*:<to.mail@web.de>:Event] evt=2
}
evt=0
```

**Reply by forumlw06** [reply]
Hello Gerhard, thanks for the quick reply.

I tried your suggestion, but unfortunately it does not work. To verify that sendmail is working in principal i slightly modified your script, so that a control mail is send every minute.

This is my script:
```
>D
t:timer=0
evt=0

>E
print --section e --
>E
if (evt==1) {
	print evt=%evt%
	=>sendmail [*:*:*:*:*:to.mail@web.de:NousA1T Event 1] evt=1
}
if (evt==2) {
	print evt=%evt%
	=>sendmail [*:*:*:*:*:to.mail@web.de:NousA1T Event 2] evt=2
}
evt=0

>S
if (timer==0) {
	print -- section S -- send mail
	=>sendmail [*:*:*:*:*:to.mail@web.de:NousA1T Section S] Mail every minute
	timer=60
}
```

And this is what i see on the ESP32 sendmail enabled Tasmota in Console (Weblog 3, on Device1 i first issued **websend [device2] script>evt=2** and then **websend [device2] script>evt=1**
```
15:53:14.951 SCR: performs "sendmail [*:*:*:*:*:to.mail@web.de:NousA1T Section S] Mail every minute"
15:53:14.963 SRC: Rule
15:53:14.964 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 67, Pld -99, Data '[*:*:*:*:*:to.mail@web.de:NousA1T Section S] Mail every minute'
15:53:32.369 RSL: RESULT = {"Sendmail":"Done"}
15:53:32.382 HTP: Konsole
15:53:51.201 HTP: Befehl
15:53:51.204 SRC: WebCommand from 162.222.7.66
15:53:51.205 CMD: Grp 0, Cmd 'SCRIPT', Idx 1, Len 6, Pld -99, Data '>evt=2'
15:53:51.208 RSL: RESULT = {"Script":">evt=2"}
15:53:51.209 --section e --
15:54:15.480 -- section S -- send mail
15:54:15.482 SCR: performs "sendmail [*:*:*:*:*:to.mail@web.de:NousA1T Section S] Mail every minute"
15:54:15.493 SRC: Rule
15:54:15.494 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 67, Pld -99, Data '[*:*:*:*:*:to.mail@web.de:NousA1T Section S] Mail every minute'
15:54:16.744 RSL: RESULT = {"Sendmail":"Done"}
15:54:19.510 HTP: Befehl
15:54:19.513 SRC: WebCommand from 162.222.7.66
15:54:19.515 CMD: Grp 0, Cmd 'SCRIPT', Idx 1, Len 6, Pld -99, Data '>evt=1'
15:54:19.518 RSL: RESULT = {"Script":">evt=1"}
15:54:19.521 --section e --
15:54:59.128 CMD: weblog
15:54:59.129 SRC: WebConsole from 54.153.56.64
15:54:59.131 CMD: Grp 0, Cmd 'WEBLOG', Idx 1, Len 0, Pld -99, Data ''
15:54:59.133 RSL: RESULT = {"WebLog":3}
15:54:59.135 --section e --
```

So looks like somehow the "evt=2" and "evt=1" variables are set, but nothing happens in the `>E` section?

Any further ideas? What do i miss?

BTW: Any idea, where these strange IP adresses in WebCommand/WebConsole logs are coming from? Both devices are on my 192.168.178.xx network.

Thanks in advance!

Regards Rainer

**Reply by gemu2015** [reply]
you have 2 >E sections. only the first is executed

**Reply by gemu2015** [reply, marked as answer]
there may only be one section >E

**Reply by forumlw06** [reply]
Autsch! That was the problem! :-(

Thanks for your help!

Regards Rainer

**Reply by forumlw06** [reply]
I have a related question:
can i set more that one variable via Websend? 

Like `websend [192.168.188.76] script>evt=1; var=2`

Semicolon does not work, Quotation neither ??

Thanks again!

**Reply by gemu2015** [reply]
semicolon should work, remove the space char

**Reply by gemu2015** [reply]
ok, space should not matter. did you define var in >D ?

**Reply by forumlw06** [reply]
you are right, space does not matter, now it seems to work with semicolon (wit or without space). Not sure, why it did not work before.

I have the var definition in >D, but maybe it was not there during my earlier tests ...

Thanks again!
